### PR TITLE
implemented nanmedian function and added it's test.

### PR DESCRIPTION
**Pull request by usama-baloch**
Hi @YushaArif99, The torch, and tensorflow backends weren't handling the nans, so we need to find another workaround for those. Implemented these two backends, other backends seem to work fine just needed some small fixes for edge cases, so those are added also.

### CODE REVIEW

File: ivy/functional/backends/tensorflow/experimental/statistical.py [hunk 1]
```diff
@@ -160,6 +161,125 @@ def corrcoef(
...
+        if tf.reduce_all(tf.math.is_nan(arr)).numpy():
+            arr = tf.reshape(arr, shape=(1, -1))
+        else:
+            arr = arr[~tf.math.is_nan(arr)]
+
+        ret = tfp.stats.percentile(
+            arr, q, axis=None, interpolation="midpoint", keepdims=keepdims
+        ).numpy()
+        if keepdims:
+            ret = set_value(ret)
```
[the remark this answers is not in the document]
Comment (usama-baloch) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:262]: you are right about the row-wise median calculation 🙂, but for high dimensional inputs like 4D, 3D. or others, it returns in the form of an array like: [12.00] or [[12.00]], so here we only need the value of `ret` which is 12.00 because if we have this `ret` in form of an array, the line :` result = tf.reshape(result, shape=shape[:-1]).numpy()` will raise an error of something can't reshape eager Tensor like. so for the safe side, this helper is needed 😅. I hope it makes it clear 🙂

File: ivy/functional/backends/tensorflow/experimental/statistical.py [hunk 2]
```diff
@@ -160,6 +161,125 @@ def corrcoef(
...
+        result.append(ret)
+
+    result = tf.reshape(result, shape=shape[:-1]).numpy()
+
+    if keepdims:
+        keepdim_shape = tuple(keepdim_shape)
+        result = tf.reshape(result, shape=keepdim_shape)
+        result = tf.cast(result, dtype)
+
+    if input.dtype in [tf.int64, tf.float64]:
```
Comment (YushaArif99) [inline ivy/functional/backends/tensorflow/experimental/statistical.py:273]: any particular reason to have these `dtype-specific` checks and type castings? Also, why not return a `tf.Tensor` using something like `tf.constant(result, dtype= input.dtype)` for example? I think `tf.constant` is generally more stable across different versions of TF and so it's better to avoid resorting to an experimental module unless there's a specific reason to do so 🙂

File: ivy/functional/backends/tensorflow/experimental/statistical.py [hunk 3]
```diff
@@ -171,20 +291,13 @@ def nanmedian(
...
-            keepdims=keepdims,
-        )
-    return tfp.stats.percentile(
-        input,
-        50.0,
-        axis=axis,
-        interpolation="midpoint",
-        keepdims=keepdims,
-    )
+        axis = tuple(axis) if isinstance(axis, list) else axis
```
Comment (YushaArif99) [inline ivy/functional/backends/tensorflow/experimental/statistical.py:289]: small point but why convert a list to a tuple only to later on convert it back to a list (inside `_nanmedian_helper`) 😅

File: ivy/functional/backends/tensorflow/experimental/statistical.py [hunk 4]
```diff
@@ -160,6 +161,125 @@ def corrcoef(
...
+            interpolation="midpoint",
+            keepdims=keepdims,
+        )
+        ret = tf.cast(ret, dtype=dtype)
+        return ret
+
+    if isinstance(axis, tuple):
+        if len(axis) == 1:
+            axis = axis[0]
+
```
[the remark this answers is not in the document]
Comment (YushaArif99) [reply ivy/functional/backends/tensorflow/experimental/statistical.py]: actually, I think we need to have this: 
```python
axis = [axis] if isinstance(axis, int) else list(axis)
```
since the previous check would cause an issue if an `int` is passed as an axis

File: ivy/functional/backends/tensorflow/experimental/statistical.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (usama-baloch) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:257]: Just a small correction: if you are specifying `keepdims=True` then this case happens which is about nested list generation from the percentile, by the definition of the keepdims means if it is true then it keeps the `last dimension` so whenever you set the `keepdims=Ture`, this case happens.
Here is an example: 
x = `[[np.nan, 12.0, np.nan, np.nan],[2., 3., 0.4, np.nan]]`
axis=`None`
keepdims=`True`

now for every iteration you print `ret`, you'll get:


arr =  tf.Tensor([2.], shape=(1,), dtype=float32)
ret =  [2.]
arr =  tf.Tensor([12.  3.], shape=(2,), dtype=float32)
ret =  [7.5]
arr =  tf.Tensor([0.4], shape=(1,), dtype=float32)
ret =  [0.4]
arr =  tf.Tensor([[nan nan]], shape=(1, 2), dtype=float32)
ret =  [[nan]

now we can't directly use `ret`, so we check if keepdims as `True` so it may be nested because of its nature then we remove its nesting. 

Now the question arises why the arr is in nested form, the answer is if you check this line: `temp = tf.reshape(temp, shape=tf.concat([tf.shape(temp)[: (dimension - len(axis))], [-1]], axis=0))`
the line is used for making the input according to our axis because we can only use axis=`None` in the `percentile` parameter and the reason is that the axis=0/1 in percentile works somehow differently than our desired functionality (which is calculating the median), we need the functionality of calculating the `median` with some fixed parameters of `percentile` so we need to make it (axis=None) here and for making it axis=None then we need to make it compatible for percentile with respect to calculating the `median`. 
if you debug the code you'll find temp's shape is changed according to the reason I explained above and that reshaped temp will be `flattened` which makes it like this: `[[ nan  2.  12.   3.   nan  0.4  nan  nan]]` with `shape=(1,8)` and then it moves further for calculation.

I hope it makes sense 🙂.

File: ivy/functional/backends/tensorflow/experimental/statistical.py (hunk 2 above)
Comment (usama-baloch) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:268]: the specific `type-casting` was to make it compatible with the `numpy` because `numpy` somehow returns the fixed types on multiple input types, so to make it compatible with `numpy`, I used this. I added these when testing multiple edge cases, so maybe something slipped from my mind, I'll test it and change it if it needs to be changed and yup, I didn't think about using `tf.constant`, I'll use this after testing it and I hope it should work fine 👍.

File: ivy/functional/backends/tensorflow/experimental/statistical.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (YushaArif99) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:255]: > x = [[np.nan, 12.0, np.nan, np.nan],[2., 3., 0.4, np.nan]]
axis=None
keepdims=True

In the case of `axis=None`, I think `nanmedian` is calculated over all the elements of the input tensor. Hence, the median would be a scalar regardless of whether `keepdims` is `True` or `False`. So the question of nested lists is irrelevant here if my understanding is correct. 
For the case when `axis` is *not* `None`, you are transposing and then reshaping the input in a form that allows you to calculate the median row-wise. Hence, regardless of the dimension of the input tensor (`4D`,`5D` etc..), you'll alqays end up with a 2D tensor. Consequently, calculating the median row-wise for a 2D tensor *should* result in a 1D output. 
And so I don see the point of having the recursive `set_value()` helper here. Hope that makes sense 🙂

[the remark this answers is not in the document]
Comment (usama-baloch) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:255]: when we have axis = `None` from the user side, we will not touch the working of `set_value` and `transposing`. the purpose of using `set_value` is for axis _not__ `None`. 
about not using recursive `set_value()` it makes sense now that it is good to not use it because we have always had 1D or 2D in `ret` and `ret` only contains `2D` when we have `nans`  else it yet returns `1D`.  we need to have multiple checks in set_value function for this.

[the remark this answers is not in the document]
Comment (YushaArif99) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:254]: > we need to have multiple checks in set_value function for this. like checking if the ret is 2D or 1D and whether it is a list or an np.ndarray I guess,

I think we can remove this function altogether and simply have `ret=ret[0]` if that makes sense 🙂

[the remark this answers is not in the document]
Comment (usama-baloch) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:254]: it only works for 1D but for 2D it returns 1D if we do so.

[the remark this answers is not in the document]
Comment (YushaArif99) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:254]: yeah but when why we have `[[nan]]` as an output 🤔 ? Can you give an example

[the remark this answers is not in the document]
Comment (usama-baloch) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:254]: there is a specific check: `if tf.reduce_all(tf.math.is_nan(arr)).numpy()` inside the jump loop wherein you are checking if all the values are nans, if they are nans then we don't need to use this line: `arr = arr[~tf.math.is_nan(arr)]` because it will remove all the `nans` but we need one nan for working on it if it make sense 🙂.

like about the above example: 
arr = tf.Tensor([2.], shape=(1,), dtype=float32)
ret = [2.]
arr = tf.Tensor([12. 3.], shape=(2,), dtype=float32)
ret = [7.5]
arr = tf.Tensor([0.4], shape=(1,), dtype=float32)
ret = [0.4]
_On this line it will not use the `tf.math.is_nan`, it just flatten it and need one `nan` because of the nature of nanmedian functionality_
arr = tf.Tensor([[nan nan]], shape=(1, 2), dtype=float32)
ret = [[nan] (this was a typo sorry for that 😅) it should be ret = [[nan]].

[the remark this answers is not in the document]
Comment (usama-baloch) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:254]: if you run this, input = `[[2.0, ivy.nan, ivy.nan, 5.6, 7.8],[ivy.nan, ivy.nan, ivy.nan,ivy.nan,ivy.nan]]`
this will produce 5.6 and [[nan]], but here we need output like: [[5.6, nan]]. this is just one case and there are multiple cases with multiple values where we need to attach them and make it specific shape output as numpy's output shape, it wouldn't be that easy to generate a same shape output when you have some values are in shape=() or in shape=(1,1) I guess so.

[the remark this answers is not in the document]
Comment (YushaArif99) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:254]: oh yeah makes sense. In that case, cant we simply do the following: 
```python
if keepdims:
   ret = tf.squeeze(ret) 
```
this will achieve the same purpose and is more intuitive I think 🙂

[the remark this answers is not in the document]
Comment (usama-baloch) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:254]: yup, good idea!
I updated the code, if anything seems incorrect let me know 🙂

File: ivy/functional/backends/tensorflow/experimental/statistical.py (hunk 2 above)
Comment (YushaArif99) [reply ivy/functional/backends/tensorflow/experimental/statistical.py:257]: Just another point regarding this, I don't think there is a need to convert the `tf.Tensors` to `np.ndarrays` throughout the code. Doing these `.numpy()` calls only adds to the computational overhead so its better to stick to native tensors 🙂

**Review by YushaArif99 — CHANGES_REQUESTED** [review]
@usama-baloch 

The implementation looks good overall. Could you also kindly add some comments in the `_nanmedian_helper` just for better code readability. Thanks! 🙂

**Comment by YushaArif99** [comment]
Hey @vedpatwardhan. Could you kindly also take a brief look at the PR? There might be a few optimizations to simplify the logic and remove any excessive loops.  Thanks! 🙂

File: ivy/functional/backends/tensorflow/experimental/statistical.py (hunk 2 above)
Comment (usama-baloch) [reply ivy/functional/backends/tensorflow/experimental/statistical.py]: I removed the `.numpy()` calls all over the function, and now it seems fine. I also updated the `dtype-casting` using `tf.cast` which looks more familiar in this scenario. About using checks there are reasons which are `numpy` always returns `float64` when we have these dtypes : `int32`, `int64`, and `float64`. if we have `float32` it returns the same.

**Comment by usama-baloch** [comment]
@YushaArif99 @vedpatwardhan, I have also added the explanation of the `_nanmedian_helper`, I tried my best to explain and concise 😅✌.

**Review by vedpatwardhan — APPROVED** [review]
lgtm! I haven't dived into the changes line-by-line, but the `nan` handling logic seems to be correct. Thanks @usama-baloch @YushaArif99 😄

**Review by YushaArif99 — APPROVED** [review]
Hey @usama-baloch 

Apologies for the late reply. Was a little swamped with work so didn't get the time to review pending PRs. 
Thanks for adding the docstring for the `nanmedian_helper`. The PR looks in good condition right now and since Ved has also approved the changes, I'll now be merging the PR. Thanks! 🙂

**Comment by YushaArif99** [comment]
> lgtm! I haven't dived into the changes line-by-line, but the `nan` handling logic seems to be correct. Thanks @usama-baloch @YushaArif99 😄

Thanks a lot for the review @vedpatwardhan! 🙂
